src/rt/aaA.d

```diff
@@ -696,18 +696,20 @@ extern (C) pure nothrow @nogc @safe
 
     bool _aaRangeEmpty(Range r)
     {
-        return r.impl is null || r.idx == r.dim;
+        return r.impl is null || r.idx >= r.dim;
     }
 
     void* _aaRangeFrontKey(Range r)
     {
+        assert(!_aaRangeEmpty(r));
```

Comment:
IMO

```d
if (_aaRangeEmpty(r))
    assert(0, "Trying to access the front of an empty associative array key range");
```

is safer and better for the user

Reply:
Like I said in the bug report, assert(0) is going to simply halt the program, and effectively generate a segfault. It's no better.

Reply:
It gives the user an error message

Reply:
https://run.dlang.io/is/wIMy1L

Reply:
Well, that sucks. It's always better to give the user an error message with the stack trace; I didn't know release just stripped that as well.

Reply:
I have a way around that, but I'm not sure it's applicable here. At this point, it's an opaque call in druntime with a `void *`. If we are to have an error, it should be in object.d I would think.

Reply:
I'm undecided on whether we should add an assert message. On one hand, I agree with Steven that currently it will not have any effect. On the other hand, assert messages do offer more information to the user about what went wrong in general, and given that Walter wants to make asserts configurable, it's not hard to imagine them implemented in terms of `core.internal.abort.abort`, when druntime is compiled in debug mode.
(If it's not already the case, it would be a good idea to make dmd use the debug version of druntime + phobos when user's code is compiled with `-debug`, I think LDC already does this.)

Reply:
The abort function is what I was referring to. I could make it abort, but I think returning `null` here is a perfectly acceptable alternative. It keeps the control of what happens in this situation out of the low-level details. I'd rather throw the error up above where we can use the type system to make a nicer error.

Reply:
+1 to I wish dmd would compile the non-release debug versions of druntime and phobos when you use the `-debug` switch. If that's not the case, it should be.